### PR TITLE
feat: await-able Async methods

### LINKED ISSUE #1515: Return Promise if callback is not provided?
Nice to have in ES2017 async/await development
```
const asyncjs = require('async');
const foo = async () => {
    const arr = Array.from({ length: 1000 }, (v, i) => i);
    const transformedArr = await asyncjs.mapLimit(arr, 100, async (v) => {
        return await Promise.resolve(v * 2);
    });
    console.log(transformedArr);
}
```

### PR DESCRIPTION
**Pull request by aearly**
Closes #1515

### CODE REVIEW

File: lib/internal/promiseCallback.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+const PROMISE_SYMBOL = Symbol('promiseCallback')
+
+function promiseCallback () {
```
Comment (aearly) [inline lib/internal/promiseCallback.js]: This is the older strategy.  It's a bit more clunky, but doesn't add another frame to the call stack for each method.

File: lib/internal/awaitify.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        return new Promise((resolve, reject) => {
+            args.push((err, ...cbArgs) => {
+                if (err) return reject(err)
+                resolve(cbArgs.length > 1 ? cbArgs : cbArgs[0])
+            })
+            asyncFn.apply(this, args)
+        })
+    }
+
+    awaitable[Symbol.toStringTag] = 'AsyncFunction'
```
Comment (aearly) [inline lib/internal/awaitify.js]: This is a bit of a hack, not sure if we should do this.

File: lib/internal/awaitify.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    awaitable[ASYNC_FN] = true
```
Comment (aearly) [inline lib/internal/awaitify.js]: I intend to use this so we can avoid an additional layer of wrapping in `wrapAsync`.

**Comment by Themayu** [comment]
> Or we just punt on this completely. You can't await wrapped functions. This would be a bummer, because `await timeout(myFn, 1000)(...args)` is pretty powerful.

How exactly would this be powerful?
(I just want an explanation because I don't really understand myself)

**Comment by aearly** [comment]
> How exactly would this be powerful?

It's powerful for the same reason async/await is powerful. `await myFn(...args)` is pretty powerful, so being able to seamlessly slip in a timeout wrapper (or memoize wrapper, or reflect) is even more powerful.

File: lib/internal/promiseCallback.js (hunk 1 above)
Comment (aearly) [reply lib/internal/promiseCallback.js]: I had to use this in a couple places where the `awaitify` wrapper couldn't work  (usually due to optional arguments).

**Comment by Themayu** [comment]
Ah. Alright then

**Comment by aearly** [comment]
@megawac @hargasinski I would love to get a review on this before merging.  In regards to my above comment about the wrapper Utils methods, I'm thinking we can defer a solution there until later (temporary punt).

**Comment by hargasinski** [comment]
@aearly really sorry for the delay. These past few weeks have busy for me, and the PR ended up getting drowned out by other things. Thanks for the reminder! I will review it this weekend. I want to spend some time playing around with it, just to make sure I understand it fully.

File: lib/internal/once.js [hunk 4]
```diff
@@ -1,8 +1,10 @@
 export default function once(fn) {
-    return function (...args) {
+    function wrapper (...args) {
         if (fn === null) return;
         var callFn = fn;
         fn = null;
         callFn.apply(this, args);
-    };
+    }
+    Object.assign(wrapper, fn)
```
Comment (megawac) [inline lib/internal/once.js]: What is the purpose of this change?

Comment (aearly) [reply lib/internal/once.js]: So static properties attached to the `fn` get attached to the `wrapper`, namely the `PROMISE_SYMBOL` attached to `promiseCallback`s.

**Comment by aearly** [comment]
@hargasinski Don't sweat it, after all, this project is run on donated time.

@megawac I hear your concerns about whether we should do this by default -- this is a rather significant change.  True, users can promisify methods, but I'm thinking about developer convenience.  Already, people are omitting the final callback on an Async method and being surprised a promise is not returned.  Several other popular libraries already work like this.

`async`/`await` is a very fundamental shift in how async code can work in JavaScript, and IMO it finally obsoletes callback-based approaches.  It feels weird to not give first-class support to it -- I feel we're somewhat obligated to because we control the package named `async` on npm.  Newer users of JavaScript are going to become more and more confused about the purpose of this library if we don't make some changes.

This next major version seems like a good point to throw more weight behind `async`/`await`-based approaches.

**Comment by megawac** [comment]
How does this affect performance Alex? If its a significant hit my vote will be to seperate the asyncified and original methods into 2 files. I'd be completely fine with the async compatible version being our default export.

I know several people who use our library due to it's performance for intensive use cases so I'm adamant about maintaining good benchmarks :)

**Comment by aearly** [comment]
There was no significant difference in the benchmarks on node 8 or 10.  I was concerned the extra wrapper would slow things down, but that is not the case.

**Comment by hargasinski** [comment]
This is some really awesome work! The change look good to me.

Personally, I'm in favour of it. I think there has been enough interest/requests (#439, #956, [promise-async](https://github.com/Kikobeats/promise-async), [koa-async](https://github.com/eladnava/koa-async), [async-q](https://github.com/eladnava/koa-async)) to justify adding it. Also, looking over past issues, it seems that historically we've been against it because promises weren't implemented natively. Now that they are, and with significant parts of the community moving towards promise-based approaches, I think this change makes sense for us.
